### PR TITLE
#4 wein empfehlung

### PR DESCRIPTION
**Pull request by Haferkorn**
MVP

### CODE REVIEW

File: backend/src/main/java/fr/anettehaferkorn/backend/controller/RecommendationController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    MatchingService matchingService;
+
+    @Autowired
+    public RecommendationController(MatchingService matchingService) {
+        this.matchingService = matchingService;
+    }
+
+    @GetMapping
+    public List<WineGrape> getAllWine(){
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/controller/RecommendationController.java:24]:
```suggestion
    public List<WineGrape> getAllWines(){
```

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Document("WineGrapes")
+public class WineGrape {
+
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java:13]: Why WineGrape and not just Wine?

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    private String name;
+    private String occasion;
+    private String wineStyle;
+    private String region;
+    private String alcohol;
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java:18]: as in alcohol content? why not a float?

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineQuery.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class WineQuery {
+
+    private String occasion;
+    private String wineStyle;
+    private String region;
+    private String alcohol;
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/model/WineQuery.java:15]: as above

File: backend/src/main/java/fr/anettehaferkorn/backend/repo/WineGrapeRepository.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.data.repository.PagingAndSortingRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface WineGrapeRepository extends PagingAndSortingRepository<WineGrape, String> {
+    public List<WineGrape> findAll();
+
+    public List<WineGrape> findWineQueryDTOByOccasionAndWineStyle(String occasion, String wineStyle);
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/repo/WineGrapeRepository.java:13]: does that work as easy as that, even with multiple attributes? Good to know. Pretty cool

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineQuery.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class WineQuery {
+
+    private String occasion;
+    private String wineStyle;
+    private String region;
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/model/WineQuery.java:14]: I think you will want to use an enum here and only allow input that your filter will be able to handle well, most probably for all attributes of this model.

File: backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+        for (RecommendationDTO match: wineMatches) {
+            if(Objects.equals(match.getRegion(), wineQuery.getRegion())||
+                    (Objects.equals(wineQuery.getRegion(),"other"))||
+                    (Objects.equals(wineQuery.getRegion(),"idK"))
+            ){
+                match.setMatchingPoints(match.getMatchingPoints()+1);
+            }
+        }
+        return matchThirdFilterLayer(wineQuery,wineMatches);
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java:49]:
```suggestion
List<String> matchers = List.of(wineQuery.getRegion(), "other", idK");
wineMatches
  .stream()
         .forEach(wine -> if(matchers.contains(wine.getRegion())) {
            wine.setMatchingPoints(match.getMatchingPoints() + 1))

        return matchThirdFilterLayer(wineQuery,wineMatches)
```

File: backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                match.setMatchingPoints(match.getMatchingPoints()+1);
+            }
+        }
+        return wineMatches;
+    }
+
+    public List<RecommendationDTO>cleanMatches(List<RecommendationDTO> matches){
+            matches.removeIf(match -> match.getMatchingPoints() <2);
+            return matches;
+    }
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java:79]: I prefer streams, but mabe it's just a matter of taste...
The filters should be named for what they're doing, i.e. filterWinesByRegion
also I wouldn't call the next Filter from each function, but rather return the value and write something like
Wines[] filteredByRegion = fitlterWinesByRegion(query, wines);
Wines[] filterByAlcoholContent = filterWinesByAlcoholContent(filteredByRegion);
...etc.
Thinking about it, they're all not real filters. 
Another possibility to do this would be to actually filter in each of these steps, join all lists together and then select/return only those who occur more than 2 times(or whatever value you choose)

File: backend/src/main/java/fr/anettehaferkorn/backend/service/QueryToRecommendationMapper.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+                    wineOfDB.getName(),
+                    wineOfDB.getOccasion(),
+                    wineOfDB.getWineStyle(),
+                    wineOfDB.getRegion(),
+                    wineOfDB.getAlcohol(),
+                    wineOfDB.getTaste(),
+                    wineOfDB.getDescription(),
+                    0));
+        }
+        return  recommendationDTOS;
```
Comment (macben888) [inline backend/src/main/java/fr/anettehaferkorn/backend/service/QueryToRecommendationMapper.java:25]: again, I would prefer streams

File: frontend/src/components/pages/recommandationForm/QuestionData.json [hunk 10]
```diff
@@ -78,40 +78,36 @@
...
       {
-        "text": "South-Europe",
+        "text": "Germany",
         "nextIndex": 5,
-        "value":"south_europe"
+        "value":"Germany"
       },
       {
-        "text": "North-America",
+        "text": "Italy",
```
Comment (macben888) [inline frontend/src/components/pages/recommandationForm/QuestionData.json:91]: There's some serious geographical confusion going on in this file

**Review by macben888 — APPROVED** [review]

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java (hunk 2 above)
Comment (Haferkorn) [reply backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java:13]: because so far these are only grape recommendations, e.g. Chardonnay, Riesling, Merlot and not explicitly wines in the sense of one of the wines XY from xyz

File: backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java (hunk 3 above)
Comment (Haferkorn) [reply backend/src/main/java/fr/anettehaferkorn/backend/model/WineGrape.java:18]: It's just light / strong

File: backend/src/main/java/fr/anettehaferkorn/backend/repo/WineGrapeRepository.java (hunk 5 above)
Comment (Haferkorn) [reply backend/src/main/java/fr/anettehaferkorn/backend/repo/WineGrapeRepository.java:13]: Yes :-)

File: backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java (hunk 8 above)
Comment (Haferkorn) [reply backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java:79]: The thought behind these three layers was that since they don't have to be set in the frontend (e.g. on I don't mind) they should not exclude entries, but only assign matchingPoints. 
Since all three have to pass through to create the ranking effect, chain them together seemed to make sense.

Comment (macben888) [reply backend/src/main/java/fr/anettehaferkorn/backend/service/MatchingService.java:79]: I can see the logic behind it. I just think it's difficult to read.
 I would definitely name the filters for what they're actually filtering for as mentioned above.
Also I think 'cleanMatches' does not state very clearly to the reader what it is actually doing. 
Also I would execute each method in the getMatchingWines method. Like this it will be immediately visible to the reader what is happening in this class
If you would just use filtered lists and then count how often each wine occurs in the combined list you would get around getting/setting matchingPoints in each filter method.

File: backend/src/test/java/fr/anettehaferkorn/backend/controller/RecommendationControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+            //Given
+            String URL = "http://localhost:" + port + "/recommendation";
+            WineQuery wineQuery= new WineQuery("dinner","red","France","strong","sweet");
+
+            //When
+            ResponseEntity<RecommendationDTO[]> response = testRestTemplate.postForEntity(URL, wineQuery, RecommendationDTO[].class);
+            //Then
+            assertThat(response.getStatusCode(), is(HttpStatus.OK));
+
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/controller/RecommendationControllerTest.java:62]: don't you also wanna compare the response.body with the result you expect to get here?

File: backend/src/test/java/fr/anettehaferkorn/backend/controller/RecommendationControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        //Given
+        String URL = "http://localhost:" + port + "/recommendation";
+        WineQuery wineQuery= new WineQuery("dinner","white","idK","light","dry");
+
+        //When
+        ResponseEntity<RecommendationDTO[]> response = testRestTemplate.postForEntity(URL, wineQuery, RecommendationDTO[].class);
+        //Then
+        assertThat(response.getStatusCode(), is(HttpStatus.OK));
+
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/controller/RecommendationControllerTest.java:85]: same as above

File: backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+        //When
+        List<RecommendationDTO> actual = matchingService.getMatchingWines(wineQuery);
+        //Then
+        assertEquals(mockedMatches, actual);
+
+    }
+
+    @Test
+    void getMatchingWines_2() {
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java:41]: find a better name that describes what this test is doing differently than the first one

File: backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+        //When
+        List<RecommendationDTO> actual = matchingService.getMatchingWines(wineQuery);
+        //Then
+        assertEquals(mockedMatches, actual);
+
+    }
+
+    @Test
+    void getMatchingWines_NoPerfectMatch() {
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java:68]: maybe add a @DisplayName that describes what you are testing here. I.e. 'returns a filtered list of wines even though there is no perfect match'

File: backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Test
+    void noMatches_Occasion() {
+        MatchingService matchingService = new MatchingService(wineGrapeRepository);
+        WineQuery wineQuery=new WineQuery("dinner","red","idK","light","sweet");
+
+        when(wineGrapeRepository.findWineQueryDTOByOccasion(wineQuery.getOccasion())).thenReturn(List.of());
+
+        //Then
+        assertThrows(NoMatchingWineException.class, () -> matchingService.getMatchingWines(wineQuery));
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java:101]: same as above

File: backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        List<WineGrape>mockedRecommendations=List.of(
+                new WineGrape("AAAA","dinner","red","France","light","sweet","abcd"),
+                new WineGrape("BBBB","dinner","red","France","strong","dry","abcd"),
+                new WineGrape("CCCC","dinner","red","France","strong","sweet","abcd"),
+                new WineGrape("DDDD","dinner","red","Italy","strong","dry","abcd"),
+                new WineGrape("EEEE","dinner","red","Spain","light","sweet","abcd")
+        );
+
+        when(wineGrapeRepository.findWineQueryDTOByOccasion(wineQuery.getOccasion())).thenReturn(mockedRecommendations);
+
```
Comment (macben888) [inline backend/src/test/java/fr/anettehaferkorn/backend/service/MatchingServiceTest.java:119]: same as above

**Review by macben888 — APPROVED** [review]
well done. I think all the filters could use a @DisplayName with a little descriptive text about what exactly the different tests are testing. Would make it easier for the reader to understand immediately the need for the different tests, and also to maybe realize if there is another edge case to be tested missing.
